### DISCUSSION TITLE
How to trigger a script when a file arrives?

### DISCUSSION THREAD

**Discussion by Yber-Aldemar** [drakkan/sftpgo#1745] [Q&A]
Hi,

My need is to trigger a script, providing the arrival directory as a parameter, when a file arrives in a directory.

There are several directories /user1/in, /user2/in, /user3/in, ... . 
User1, user2, ... connect to SFTPGo to upload files.

When a file arrives, from these users & in one of these directories, we must trigger a script new_file.ksh, with the parameter of the directory in which there is a file that has just arrived.
I tried putting the full path in the event rule, but nothing happens.
The Event action is OK, when i launch it with an Event rule on demand, the script is OK.
I think I don't understand how it works with the trigger Filesystem event.
My tests are done with version 2.5.6.
Can you help me?

Thanks in advance :)
Regards

PS: sorry for my English, it's not my native language.

**Comment by nzx-jsantaana** [comment]
When you're setting up an action like this, there are three important things to consider:

1. you set up the parameters in the Action definition (not the Rule); by default, a Command-type Action will not be passed any parameters.
2. you need to use the _right_ parameters: eg, {{VirtualPath}} vs {{FsPath}}
3. the path filters in the Rule are the _virtual_ path (which may vary by user), not the "real" FS path (because some storage backends might not have a "real" directory-type path)

We have a similar case of "trigger an action on file upload", and we use the following setup.

_Action_ definition:
* Command: path to the script on the server
* Arguments: {{Name}},{{ObjectName}},{{VirtualDirPath}},{{TargetName}}
We work out the real path based on the first three, because we need to preserve parts of the VirtualPath and it's easier this way; you might find it easier to just pass {{FsPath}}.  There's a reason for the {{TargetName}}, but we'll get to that later.

_Rule_ definition:
* Trigger: Filesystem events
* Fs events: upload, rename
* Name filters: one entry, "*" (the default, I think)
* Group name filters: an entry for each user group we want the trigger to apply to
* Path filters: an entry for each directory we want to watch, ending with "**"
* Actions: one entry, the Action we defined previously

You could use the Name filters instead of Group if you prefer; we find groups to be tidier but it doesn't make much difference if you only have a few users.  For your case, the Path filters would depend on whether users have isolated virtual dir setups or not.

If they do (ie, user1 uploads to "/in", and user2 also uploads to "/in", but they're not the same directory) then you just need one Path filter: "/in/**".  (It might be trickier if you let users create subdirectories, but we don't.)

If they all see the same dir structure (user1 knows they're uploading to "/user1/in", user2 knows they're uploading to "/user2/in", etc) then you can either list them individually (if you only have a few users), or use a wildcard like "/user*/in/**".  Obviously using a wildcard in the path risks matching a wider set than you expected, so be careful.

Other things to be careful of:
1. if you typo a user/group name and it doesn't match anything, SFTPGo won't warn you, so double-check.
2. some SFTP clients will upload files under a temporary name (ie "file.ext.part"), then rename them (to "file.ext") when the transfer is complete.  This is so they can resume interrupted uploads, but it's a problem for on-upload triggers.  You can set upload_mode=1 to avoid this, but it won't change client behaviour, so we just have our processing action wait a few seconds (to avoid a race with the client rename), and trigger on rename as well so we get the new name of the file (which is why we have {{TargetName}} in the Action parameters).

**Comment by Yber-Aldemar** [comment]
Hello,

Thank you very much, this helped me a lot. 
I couldn't correctly specify the paths to monitor, and your answer gave me the elements I was missing.
Thanks again :)

**Comment by sliprealm** [comment]
Hi everone, 

Thank you all for the explanation, following your tips i managed to trigger a python scrip with success, however whenever i try to bring up a container that is actually called from a bash script with parameters being sent from the server side,  i get an error as described below:
Parameters
`{{Event}},{{FsPath}},{{FileSize}},{{IP}},{{Name}},{{ObjectName}}`

Bash script (test) 
`/usr/bin/docker run \
-v /srv/sftpgo:/srv/sftpgo \
-it vydra-pinger:latest $1 $2 $3 $4 $5 $6 $7
`
Error log: 
`{"level":"debug","time":"2024-09-25T08:54:29.999","sender":"eventmanager","message":"executed command \"/srv/scripts/vydra_pinger/docker.sh\", elapsed: 22.139488ms, error: exit status 1"}
{"level":"error","time":"2024-09-25T08:54:29.999","sender":"eventmanager","message":"unable to execute action \"VydraPinger-OnFileUpload\" for rule \"VydraPinger-OnFileUpload\", elapsed 22.262857ms, err: action \"VydraPinger-OnFileUpload\" failed: exit status 1"}
`
If I exclude calling of docker, it works ok, so is suspect that must be something on a system level when calling docker (permissions?). However i put sftpgo user into docker group, but still without success. 

Anyone out there that is calling docker containers from events /  actions had simillar issues? 

Many thanks,

Seb

**Reply by sliprealm** [reply]
Anyone out there that could give it a try and help me with this ? Much appreciated.
